stateresolution.go

```diff
@@ -0,0 +1,255 @@
+	for _, block := range r.others {
+		if event := r.resolveNormalBlock(block); event != nil {
+			r.result = append(r.result, *event)
+		}
+	}
+	return r.result
+}
+
+// A stateResolver tracks the internal state of the state resolution algorithm
+// It has 3 sections:
+//
+//  * Lists of lists of events to resolve grouped by event type and state key.
+//  * The resolved auth events grouped by type and state key.
+//  * A List of resolved events.
+//
+// It implements the AuthEvents interface and can be used for running auth checks.
+type stateResolver struct {
+	creates                   []Event
+	powerLevels               []Event
+	joinRules                 []Event
+	thirdPartyInvites         [][]Event
+	members                   [][]Event
+	others                    [][]Event
+	resolvedCreate            *Event
+	resolvedPowerLevels       *Event
+	resolvedJoinRules         *Event
+	resolvedThirdPartyInvites map[string]*Event
+	resolvedMembers           map[string]*Event
+	result                    []Event
+}
+
+func (r *stateResolver) Create() (*Event, error) {
+	return r.resolvedCreate, nil
+}
+
+func (r *stateResolver) PowerLevels() (*Event, error) {
+	return r.resolvedPowerLevels, nil
+}
+
+func (r *stateResolver) JoinRules() (*Event, error) {
+	return r.resolvedJoinRules, nil
+}
+
+func (r *stateResolver) ThirdPartyInvite(key string) (*Event, error) {
+	return r.resolvedThirdPartyInvites[key], nil
+}
+
+func (r *stateResolver) Member(key string) (*Event, error) {
+	return r.resolvedMembers[key], nil
+}
+
+func (r *stateResolver) addConflicted(events []Event) {
+	type conflictKey struct {
+		eventType string
+		stateKey  string
+	}
+	offsets := map[conflictKey]int{}
+	// Split up the conflicted events into blocks with the same state key.
+	for _, event := range events {
+		key := conflictKey{event.Type(), *event.StateKey()}
+		// Work out which block to add the event to.
+		var block *[]Event
+		// By default we add the event to a block in the others list.
+		blockList := &r.others
+		switch key.eventType {
+		case "m.room.create":
+			if key.stateKey == "" {
+				block = &r.creates
+			}
+		case "m.room.power_levels":
+			if key.stateKey == "" {
+				block = &r.powerLevels
+			}
+		case "m.room.join_rules":
+			if key.stateKey == "" {
+				block = &r.joinRules
+			}
+		case "m.room.member":
+			blockList = &r.members
+		case "m.room.third_party_invite":
+			blockList = &r.thirdPartyInvites
+		}
+		if block == nil {
+			// We need to find an entry for the state key in a block list.
+			offset, ok := offsets[key]
+			if !ok {
+				// This is the first time we've seen that state key so we add a
+				// new block to the block list.
+				offset = len(*blockList)
+				*blockList = append(*blockList, nil)
+			}
+			// Get the address of the block in the block list.
+			block = &(*blockList)[offset]
+		}
+		// Add the event to the block.
+		*block = append(*block, event)
+	}
+}
+
+// Add an event to the resolved auth events.
```

Comment:
I don't understand the significance of the term "resolved". AFAIK, auth events are a select number of state event types (the ones inside the switch in the body of this function) and that's it. There is no concept of them being "resolved" or "unresolved" or anything else. Indeed, grepping for `unresolved` turns up nothing, so what does this mean?